### ISSUE TITLE
Improve sanitization for list of URI types

### ISSUE THREAD

**Issue by crmky** [spring-projects/spring-boot#23037]
**Problem**

I'm using Spring Boot 2.2.9. The change introduced by https://github.com/spring-projects/spring-boot/pull/19999 considers any keys contains "uri", "uris", "address" or "addresses" are "comma separated URLs". This is not always the right assumption. It will try to remove password from those URLs, however if it's not a URL format, it will **return the original content**. 

**Expected Behavior**

If that key is not a URL format, it should return as `******`. Or at least allow developers to configure whether they want sanitize URLs or not.

**Reproducer**

```java
Sanitizer sanitizer = new Sanitizer();
System.out.println(sanitizer.sanitize("uris", "[amqp://foo:bar@host/]"));
System.out.println(sanitizer.sanitize("uris", "amqp://foo:bar@host/"));
```

The output is:
```
[amqp://foo:bar@host/]
amqp://foo:******@host/
```

**Comment by bclozel** [comment]
Thanks for raising this issue.
We agree that the behavior is not consistent with our documentation and the whole content should be sanitized if the property is not of `URI` type.

**Comment by helloworldless** [comment]
This looks reasonable for my first contribution! Taking a look...

**Comment by Davio** [comment]
In this case, the part inside the square brackets is still a normal URL which would be properly sanitized if parsed without the square brackets. Wouldn't it make more sense to make the regex more flexible so it can detect embedded URLs?

For instance, changing it to `.*[A-Za-z]+:/.+:(.*)@.+` ?

**Comment by helloworldless** [comment]
That solution definitely makes sense as well. I just implemented the fix in this way per @bclozel's suggestion which I hope I've understood correctly:

> sanitize the password section only if the value matches the user info pattern, or completely sanitize the value and return ******

I was able to justify this solution to myself as follows...

There are probably other patterns out there too, like the URI surrounded by curly braces instead of square brackets, or some other format we can't even imagine. If we accept that it's unlikely that we would ever be able to account for every possible pattern, we can make things simple, and just sanitize anything that doesn't match the most common pattern. That will undoubtedly cause us to sanitize values which do not contain passwords, but it also will prevent us from accidentally exposing a password which is contained in an unhandled pattern.

**Comment by Davio** [comment]
I certainly understand the route you have chosen. It's better to err on the side of caution especially when security is concerned.
Do you know the purpose behind the square brackets for AMQP addresses?

We also use AMQP and specify the username and password separately so it is not embedded in the URL.

**Comment by bclozel** [comment]
@helloworldless Sorry my first comment was wrong - we've discussed that during a team call and this issue was more subtle than we thought. I've updated the issue description and the steps to follow. Hopefully now everything is clear.

Given the unfortunate back and forth on this issue, feel free to request help or to leave out part of the work - we'll happily complete it before merging.

Thanks!

**Comment by helloworldless** [comment]
No worries. I'll take another look!

**Comment by helloworldless** [comment]
After a bit of experimentation, it looks like it's going to be very tricky to come up with a single regex pattern that works for lists and individual values. So what if we handle an individual value and a list differently? An individual value would be handled basically like it is currently, and a list (identified by surrounding square brackets) would be handled something like this:

1. Strip the square brackets
1. Split the elements on commas
1. Apply regex and replacement to each element
1. Rejoin the elements with commas and surrounding square brackets

**Comment by bclozel** [comment]
@helloworldless wouldn't the regexp proposed in the issue description (i.e. `"\\[?[A-Za-z]+://.+:(.*)@.+$"`) work in this case? If I'm not mistaken we already split the value on commas and then match against the pattern. An optional `[` in the pattern should cover our missing case?

**Comment by helloworldless** [comment]
I spent some more time on this, and here's the pattern I came up with: 

`String proposedPattern = "(\\[?[A-Za-z]+?:\\/\\/.+?:)(.*?)(@.+?)";`

I ended up needing the three capture groups to be able reconstruct the values. I also had to make the matchers lazy ("ungreedy").

I've demonstrated this here: https://repl.it/@helloworldless/spring-boot-issue-23037. This is also useful to see the regex dissected: https://regex101.com/r/sr6uFN/1.

Shall I proceed with this?

**Comment by bclozel** [comment]
I'm probably missing something here. We don't run the pattern on the whole string, since we're splitting on `","` in the first place.

I've locally changed the pattern to `Pattern.compile("\\[?[A-Za-z]+://.+:(.*)@.+$");` and ran the following test successfully:

```
	@ParameterizedTest(name = "key = {0}")
	@MethodSource("matchingUriUserInfoKeys")
	void uriAsListWithPasswordShouldHaveThoseSanitized(String key) {
		Sanitizer sanitizer = new Sanitizer();
		assertThat(sanitizer.sanitize(key,
				"[http://user1:password1@localhost:8080,http://user2@localhost:8082,http://localhost:8083]")).isEqualTo(
				"[http://user1:******@localhost:8080,http://user2@localhost:8082,http://localhost:8083]");
	}
```

Am I missing a test case that would invalidate this approach?

**Comment by helloworldless** [comment]
My bad. I had not properly investigated the code.

I have now used the pattern you suggested and updated the tests. I also modified `Sanitizer#getPattern` to handle indexed property keys like `uris[1]`. See diff here: https://github.com/spring-projects/spring-boot/compare/2.2.x...helloworldless:23037-sanitize-uris.

I do have one question about the change to `Sanitizer#getPattern`. My change alters the behavior for all sensitive keys, not just the URI/address ones. So for example, it will now match "my.password[0]", but before that would not have been a match. Is this acceptable or should I do a bit of refactoring to split out the sensitive keys (e.g. "token", "password") from the URI/address keys (e.g. "uri", "address"), so that they can be treated differently in `Sanitizer#getPattern`?

I also wanted to give a bit more detail based on my current understanding of the requirements to make sure we're on the same page...

There are actually two property list scenarios which are now handled: **"Native" Properties List** and **User-Provided List Literals**.

**"Native" Properties List**

```text
uris[0]=http://user1:password1@localhost:8080
uris[1]=http://user2:password2@localhost:8082
```
```yaml
uris:
  - http://user1:password1@localhost:8080
  - http://user2:password2@localhost:8082
```

`Sanitizer#sanitize` is called twice with these parameters:

1. key = "uris[0]", value = "http://user1:password1@localhost:8080"
1. key = "uris[1]", value = "http://user2:password2@localhost:8082"

Previously, this was not handled properly due to the trailing `value + $` in this pattern in `Sanitizer#getPattern`:

`Pattern.compile(".*" + value + "$", Pattern.CASE_INSENSITIVE)`

To fix this I've modified that pattern to match an optional index enclosed in square brackets:

`Pattern.compile(".*" + value + "(?:\\[.+])?$", Pattern.CASE_INSENSITIVE)`

**User-Provided List Literals**

I'm not sure why this would be used, but I suppose some libraries may be expecting this format.

```text
uris=[http://user1:password1@localhost:8080,http://user2:password2@localhost:8082]
```

**Note:** Quotes here are required or it's parsed as a "native" list
```yaml
uris: "[http://user1:password1@localhost:8080,http://user2:password2@localhost:8082]"
```

For either of these, `Sanitizer#sanitize` is called once with these parameters:

1. key = "uris", value = "[http://user1:password1@localhost:8080,http://user2:password2@localhost:8082]"

As you pointed out, `Sanitizer#sanitizeUris` splits on ",". So `Sanitizer#sanitizeUri` is called twice:

1. "[http://user1:password1@localhost:8080"
    1. This is now matched as a URI because of the change to `Sanitizer#URI_USERINFO_PATTERN`
1. "http://user2:password2@localhost:8082]" 
    1. This was never a problem; it was already being matched by the existing `Sanitizer#URI_USERINFO_PATTERN`

**Comment by bclozel** [comment]
Thanks for your first contribution @helloworldless - I took your commit and just simplified it a bit.

I think this first-timer issue was more subtle than expected, but the fix looks right. I ended up removing the more complex parts about matching array-like keys, because I think there was a confusion about that. The `Sanitizer` is only involved in the Actuator endpoints, when iterating on the parsed `PropertySource` instances; so the Sanitizer is never looking at the properties/yaml syntax, but rather the parsed representation of keys (`"[http://user1:password1@localhost:8080,http://user2:password2@localhost:8082]"` being the string representation of a list here).

This is now merged in 2.2.x and merged-forward in later branches.
